WP.com Plans: Result screen

guarani opened the pull request:
Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/20690

Integrates the final purchase result screen into the WP.com plan purchase flow.

## Visuals

| New screen | Entire flow |
| - | - |
| <img alt="Screenshot 2023-06-01 at 18 43 07" width="400" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/370f512b-23da-437d-88b1-69f949a1208b"> | <video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/eca98d21-edc0-4022-b3ff-66995f81b73f"> |





## Testing steps

On iPhone:

1. Log in using a WP.com account
2. Enable `Free to Paid Plans Dashboard Card` feature flag
3. Tap the `Free domain with an annual plan` card
4. Follow the steps to purchase a WP.com plan 
5. Expect to see the new "All ready to go!" screen at the end with the domain name in bold
6. Tap done to close the flow

Known issues:
- https://github.com/wordpress-mobile/WordPress-iOS/issues/20789
- https://github.com/wordpress-mobile/WordPress-iOS/issues/20790
- ~~https://github.com/wordpress-mobile/WordPress-iOS/issues/20791~~

I think these can be handled in a follow-up PR so I created separate issues for them, but happy to reconsider if you think they should be done on this PR.

## Regression Notes

1. Potential unintended areas of impact

This PR could introduce bugs in the plan purchase flow, but nothing outside of it.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I manually tested the flow as described above

3. What automated tests I added (or what prevented me from doing so)

The UI tests will be updated as mentioned above.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] VoiceOver.
- [x] ~~Languages with large words or with letters/accents not frequently used in English.~~ ⚠️ I only tested that text overflowed correctly, I don't expect issues with non-plain text because of built-in UTF8 support. 
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [x] Multi-tasking: Split view and Slide over. (iPad)

Commit:
Adds the WP.com plan purchase result screen

* Adds the screen
* Fires an analytics event when the screen is shown

staskus commented:
Regarding known issues,

- I think https://github.com/wordpress-mobile/WordPress-iOS/issues/20789 could be part of this PR, since this PR builds the UI. The way it's solved is by observing `traitCollectionDidChange` and rebuilding attributed string. (See `StatsTotalInsightsCell` for example.)

- I agree https://github.com/wordpress-mobile/WordPress-iOS/issues/20790 is outside the PR. I haven't' seen UI tests mocking web views in some ways, not sure if current mocking service even supports it, so it can have complications.

- https://github.com/wordpress-mobile/WordPress-iOS/issues/20791 could also be done in the scope of this PR. I think calling `blogService.syncBlogAndAllMetadata(blog) { }` before presenting the success view should be enough. (See `DomainsDashboardFactory`)

staskus commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Free to Paid Plans/FreeToPaidPlansCoordinator.swift:
```diff
@@ -14,26 +15,41 @@ import UIKit
 
-        let planSelected = { checkoutURL in
+        let purchaseCallback = { (domainName: String) in
+            let resultView = DomainResultView(domain: domainName) {
```
I think calling `blogService.syncBlogAndAllMetadata(blog) { }` will synchronize the blog and hide the dashboard card

guarani replied:
Makes sense! Fixed in 86bf8f1fa6df36cd0fe69c7ad27fc042fb45ec47.

staskus commented on WordPress/Classes/ViewRelated/Domains/Views/DomainResultView.swift:
```diff
@@ -0,0 +1,114 @@
+
+    var body: some View {
+        ZStack {
+            VStack(alignment: .leading, spacing: Metrics.interItemVerticalPadding) {
```
For smaller screens, horizontal orientation, or larger types in requires to have a scroll view. Similar to what was already done for `DomainCreditRedemptionSuccessViewController`

![Simulator Screen Shot - iPhone 14 Pro - 2023-06-02 at 11 58 02](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/d0a7ce36-b246-45fc-bece-f68d1d822b58)

guarani replied:
Good call! I thought this screen flow was portrait only but must've had rotation-lock on. Fixed in e66b39bec236e400def8ede01d9f528974a0c9a9.

staskus commented on WordPress/Classes/ViewRelated/Domains/Views/DomainResultView.swift:
```diff
@@ -0,0 +1,114 @@
+        var configuration: (UIViewType) -> Void
+    }
+
+    var body: some View {
```
Changing orientation makes the site name disappear. I suppose it can be fixed with the same solution to fix attributed string + dynamic type, by observing trait collection changes

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/d1630588-669a-4786-916d-7c609f7fd0da

guarani replied:
Thanks for the suggestion! 
I can listen to trait changes in SwiftUI through code such as:
```
@Environment(\.sizeCategory) var sizeCategory
```
but I'm not sure how to refresh the view based on the size category change.

So I tried a different approach using `GeometryReader` to get the current screen width and use that in the `sizeThatFits` (instead of using the label's bounds like I did before). This works somewhat because the label now is sized correctly based on the current dynamic type size and device orientation. However, it doesn't respond to changes to dynamic type size after this screen is already displayed.

I looked through `StatsTotalInsightsCell` for inspiration but the missing piece for me is how to make the size category change re-draw the screen.

staskus commented on WordPress/Classes/ViewRelated/Domains/Views/DomainResultView.swift:
```diff
@@ -0,0 +1,114 @@
+
+                AttributedLabel(dynamicHeight: $subtitleHeight) {
+                    $0.attributedText = subtitleWith(domain: domain)
+                }.foregroundColor(Color(UIColor.muriel(color: .text)))
```
Code formatting could be improved by aligning all attributes:

```
HStack {
}
.attribute()
.attribute()
```

instead of:
```
HStack {
}.attribute()
.attribute()
```

guarani replied:
Good call! Fixed

staskus commented on WordPress/Classes/ViewRelated/Domains/Views/DomainResultView.swift:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    private enum Metrics {
+        static let screenPadding = 15.0
```
Nitpick. The padding should be `16`. Generally, it's a multiple of 4 (or 2).

<img width="490" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/f63583d6-d59a-498e-986a-c3ce694752e7">

You can check other metrics as well, they are all likely to be multiple of 4s, including the corner radius.

guarani replied:
I'm glad you flagged this, I went back and updated the spacings and sizes to match the Figma design.

staskus reviewed (commented):
Thanks for the work!

I made a couple of comments and suggestions! I'll come back to review it on Monday.

staskus commented on WordPress/Classes/ViewRelated/Domains/Views/DomainResultView.swift:
```diff
@@ -0,0 +1,114 @@
+                                              value: "It may take up to 30 minutes for your domain to start working properly",
+                                              comment: "Notice on the domain purchase result screen. Tells user how long it might take for their domain to be ready.")
+        static let primaryButtonTitle = NSLocalizedString("freeToPaidPlans.resultView.done",
+                                                          value: "Done!",
```
I think "!" is not needed.
```suggestion
                                                          value: "Done",
```

Commit:
Fix SwiftUI code formatting

Commit:
Fix code formatting indent

Commit:
Fix typo in button copy

Commit:
Visual tweak to match designs

Commit:
Made domain purchase screen scrollable

Commit:
Fix cut-off text on domain purchase screen

Commit:
Refresh the dashboard after plan purchase

guarani commented:
Hey @staskus, thanks a lot for the review. I've gone and addressed these points, the only one outstanding is the dynamic type support which is improved but still not 100%. 

It's ready for another review because I think the outstanding issue is not a blocker yet (but I'd love to fix it).

staskus reviewed (approved):
👍

Thank you for the changes! It all works well now.

Regarding the dynamic type, I think most importantly the label appears in an expected size when the view is opened. I don't think it's necessary for the success view to respond to the dynamic changes when the view is opened, especially given the temporary nature of the view. We can implement it separately if we find it necessary.

Commit:
Merge branch 'trunk' into task/plans-success-view

guarani commented:
> Regarding the dynamic type, I think most importantly the label appears in an expected size when the view is opened. I don't think it's necessary for the success view to respond to the dynamic changes when the view is opened, especially given the temporary nature of the view. We can implement it separately if we find it necessary.

Sounds good! Thanks for the review 🙇 I resolved the conflict and enabled auto-merge.
